x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/EventEmitterService.java

```diff
@@ -0,0 +1,80 @@
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.Marker;
+import org.apache.logging.log4j.MarkerManager;
+import org.elasticsearch.ResourceNotFoundException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.common.io.stream.ReleasableBytesStreamOutput;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentFactory;
+
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+
+public class EventEmitterService {
+    private static final Logger logger = LogManager.getLogger(EventEmitterService.class);
+    private static final Marker AUDIT_MARKER = MarkerManager.getMarker("org.elasticsearch.xpack.application.analytics");
+    private final AnalyticsCollectionResolver analyticsCollectionResolver;
+    private final ClusterState clusterState;
+    private final BigArrays bigArrays;
+
+    public static final ParseField EVENT_DATASET_NAME_FIELD = new ParseField("event.dataset");
+    public static final ParseField TIMESTAMP_FIELD = new ParseField("@timestamp");
+
+    public EventEmitterService(AnalyticsCollectionResolver analyticsCollectionResolver, ClusterState clusterState, BigArrays bigArrays) {
+        this.analyticsCollectionResolver = analyticsCollectionResolver;
+        this.clusterState = clusterState;
+        this.bigArrays = bigArrays;
+    }
+
+    /**
+     * Emits an {@link AnalyticsEvent} to the appropriate collection.
+     *
+     * @param event The {@link AnalyticsEvent} to emit.
+     * @param listener The {@link ActionListener} to notify when the event has been emitted.
+     */
+    public void emitEvent(final AnalyticsEvent event, final ActionListener<AcknowledgedResponse> listener) {
+        final String collectionName = getAnalyticsEventCollectionName(event);
+        final AnalyticsCollection collection = analyticsCollectionResolver.collection(clusterState, collectionName);
+
+        if (collection == null) {
+            listener.onFailure(new ResourceNotFoundException(collectionName));
+            return;
+        }
+        try (ReleasableBytesStreamOutput buffer = new ReleasableBytesStreamOutput(0, bigArrays.withCircuitBreaking())) {
+            try (XContentBuilder source = XContentFactory.jsonBuilder(buffer)) {
+                source.startObject()
+                    .field(EVENT_DATASET_NAME_FIELD.getPreferredName(), collection.getEventDataStream())
+                    .field(TIMESTAMP_FIELD.getPreferredName(), getTimestamp())
+                    .endObject();
+                logger.info(AUDIT_MARKER, event.toXContent(source, null));
+                listener.onResponse(AcknowledgedResponse.TRUE);
+            }
```

Comment:
We shouldn't log the json string serialisation right away. As explained above we should log through the `ESLogMessage` class and let log4j deal with the xcontent representation.